### ISSUE TITLE
Use ABP (with Module-Zero) as a Authorization Server

### ISSUE THREAD

**Issue by GutierrezDev** [aspnetboilerplate/module-zero-forsaken#387]
Hi guys. I have been searching for a way of doing some integration between existing projects and ABP functionalities. 
All the samples that I see have module-zero inside but what I need is something like this:

ProjectA: Security Manager
*This one implement everyting module-zero offers: Users, Roles, Permission Check, etc.
*Also Login, User registration, etc.

Project B, 
*Project without module-zero but authenticate with ProjectA with provided username, password, etc.
*Permission Check with the ones granted in ProjectA.
**So here by example I want to use something like this: 
[AbpMvcAuthorize(PermissionsName.UsersEdit)]
public void UserEdit(UserEditDto input)
{

}

Project C 
*Same project B

So my problem is that I can't find a way to do it without having all the things that module-zero has (DbContext, User, Tenant,etc tables).

In other words: create an application as a central authentication and authorization end-point and use the same feature. Hope you understand :)

**Comment by wocar** [comment]
Yes, you can integrate identityserver3 and use oauth it is possible to do
this, I've done it.

**Comment by GutierrezDev** [comment]
Thanks for replying.
Do you have an example?

**Comment by hikalkan** [comment]
If you are using asp.net core then we have an integration with identity server 4 (https://aspnetboilerplate.com/Pages/Documents/Zero/Identity-Server).

**Comment by GutierrezDev** [comment]
Thanks @hikalkan. 
I dont have problem using the asp.net core version but my intention is to use multiple applications with one set of user accounts. 
So in ProjectA I manage users, tenants, roles, permissions, etc.
And in other projects (that have its own database and don't have Users table) I log in with the users created in ProjectA.

**Comment by hikalkan** [comment]
Identity Server 4 is for that: You can use one of the applications (say "main" app) as user source and make other applications login with the main application. But permissions can not be stored in the main app because every application know their own permissions (I think this should be expected). And also, other applications will have Users tables but those tables will just used to store user informations locally, they will not use their own Users tables for authentication.

But we don't provide any guide or documentation to implement this scenario, because this is completely related to Identity Server and it has their own documentation for it: https://identityserver4.readthedocs.io/en/release/quickstarts/3_interactive_login.html

**Comment by GutierrezDev** [comment]
Thanks again with your support. 
"But permissions can not be stored in the main app because every application know their own permissions (I think this should be expected)" The problem doing this way for what I want to do is that even I have every application having its own permissions, managing would be too difficult if I don't have it in a centric application. 
But I think we can make this work doing something like this: 
In other applications when you create permissions you save it to the central database through "main App" Permission service. 
When you login, you get associated roles and permissions (hope I can do it this way). What do you think @hikalkan ?

**Comment by wocar** [comment]
Identity server is an identity server not a permission server. I think each
app/service should manage permissions in its own way otherwise it would be
a nightmare to centrally manage te permissions based on roles/claims.

I'm still in the process of figuring out, if this is the best approach

**Comment by GutierrezDev** [comment]
@wocar I agree.
I was doing some research and found a demo of something what I want:
https://github.com/codingfreak/blogsamples/tree/master/AspNetIdentity
It implements asp.net identity as a service.
Hope to verify if this can satisfy my need. Thanks again.

**Comment by wocar** [comment]
Identity server and Module Zero are already based on this framework! If you
already have a user table just run identityserver pointing to your existing
database and it should work. However in my case because there is not
identityserver3 nuget and I'm not using net core I had to setup all the
registrations manually but now, it's working.

**Comment by hikalkan** [comment]
In addition;

ABP and module-zero has good extension points. Example;

* [IPermissionChecker](https://github.com/aspnetboilerplate/aspnetboilerplate/blob/dev/src/Abp/Authorization/IPermissionChecker.cs) is the main extension point that you can check if a user has granted for a permission. You can replace this by your own implementation to check permissions from any source. Module-zero [implements](https://github.com/aspnetboilerplate/module-zero/blob/dev/src/Abp.ZeroCore/Authorization/PermissionChecker.cs) this to check from UserManager.
* Another extension point can be replacing ``IRepository<UserPermissionSetting, long>`` with your own implementation that works with a second dbcontext connects to another database. Thus, not only 'checking permissions' works but also changing permissions will properly work.

You can investigate source code to find proper ways of extending it based on your needs.

**Comment by GutierrezDev** [comment]
Wow thanks again for the support you give to us.
I was thinking in using another dbContext that have the user information and permissions. i will check how to override the default behavior.

**Comment by wocar** [comment]
Use oauth in conjunction with identityserver!! You will not regret that.
And regarding permissions, te approach I am using is very basic. I use the
settings manager for permissions and I wire it up using pernissionchecker

**Comment by GutierrezDev** [comment]
@wocar Do you have a repo with that integration?
